main/src/library/ChannelImpl.flix

```diff
@@ -0,0 +1,379 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+namespace ChannelImpl {
+
+    ///
+    /// The tuple corresponds to (
+    ///   `id`,
+    ///   `channelLock` - the lock of the channel,
+    ///   `buffered` - indicates whether a handoff is needed or not.
+    ///       If you try to put an element in a channel that's full, you wait until there's space,
+    ///   `maxQueueSize` - the maximum number of elements allowed in `deque`. This is `max(1, channel size)`,
+    ///   `deque` - a deque of the channel elements,
+    ///   `waitingGetters` - a set of conditions that are waiting for an element.
+    ///       This set is cleared after each new element,
+    ///   `waitingSetters` - a condition that can notify threads of available space in `deque`.
+    /// )
+    /// The reason for the difference between `waitingGetters` and `waitingSetters` is the select expression.
+    /// A thread can only wait for one condition which is a problem when a select expression requires waiting for a
+    /// set of channels. Therefore the thread makes a condition that represents the select expression and hands that to
+    /// every channel involved.
+    ///
+    /// put is not supported in select so a single condition can be reused for all insertions.
+    ///
+    pub enum ChannelImpl {
+        case Mpmc(
+                Int64,
+                ReentrantLock,
+                Bool,
+                Int,
+                MutDeque[Boxed],
+                MutList[(ReentrantLock, Condition)],
+                Condition
+            )
+    }
+
+    ///
+    /// Creates a new channel if `bufferSize` is not negative.
+    ///
+    pub def new(bufferSize: Int): Result[ChannelImpl, String] & Impure =
+        if (bufferSize < 0) Err("Channel's buffer size must be non-negative.")
+        else Ok(unsafeNew(bufferSize))
+
+    ///
+    /// Creates a new channel. A runtime error occurs if `bufferSize` is negative.
+    ///
+    pub def unsafeNew(bufferSize: Int): ChannelImpl & Impure =
+        import dev.flix.runtime.GlobalCounter:newId();
+        let _bufferCheck = if (bufferSize < 0) inputBug("newChannelWithQueue", "bufferSize < 0") else ();
+        let buffered = bufferSize == 0;
+        let reentrantLock = ReentrantLock.new(false);
+        Mpmc(
+            newId(),
+            reentrantLock,
+            buffered,
+            if (buffered) 1 else bufferSize,
+            MutDeque.new(),
+            MutList.new(),
+            ReentrantLock.newCondition(reentrantLock)
+        )
+
+    ///
+    /// Sends the element `boxedElement` on the channel `channel`.
+    ///
+    /// Implements the expression `channel <- x`.
+    ///
+    pub def put(channel: ChannelImpl, boxedElement: Boxed): ChannelImpl & Impure =
```

Comment:
IIRC `STYLE.md` suggests (or should suggest) that we use single variable names, e.g. `c` and `e` here.

Reply:
Is single letter variables preferred for everything? That seems controversial?

Reply:
Its standard in FP languages.. because the types hold a lot more of information.